Feature: Prompt manager (cherrypicked onto stage)

artisticMink opened the pull request:
This MR affects chatcompletion APIs. 

### Features

- Change the order in which prompts are sent for each character.
- Append, remove, edit, enable or disable prompts.
- Export and import prompts either globally or per character.
- Token counts displayed for each prompt as well as accumulated tokens.

Editing 'utility' prompts like start new chat, start example chat, continue nudge, etc.

### What changed

- Introduces 'PromptManagerModule'. This module takes care of user interactions with chatcompletion settings.
- Introduces 'ChatCompletion'. It's a list that operates on a token budget. Allowing the list to be filled until the budget is spent. 
- Reworks a significant part of openai.js. Especially the prepareOpenAIMessages(...) function.
- Adds a couple events in multiple places.
- Adds the placeholders {{group}} and {{groupIfNotChar}} to substituteParams(..)
- Adds a configuration migration, moving main prompt, nsfw prompt and jailbreak prompt to the prompt manager scope and deletes the properties afterwards.

### Testing

PromptManager logging can be enabled by checking 'Log prompts to console' in user settings. 

**API Connections**

- Changing API while chatting does not cause interruption. :heavy_check_mark:
- Changing ChatCompletion Source while chatting does not cause interruption.  :heavy_check_mark:
- Changing Models while chatting does not cause interruption.  :heavy_check_mark:

**Configuration**

- Prompts and their order are persisted in presets. :heavy_check_mark:
- Prompts and their order are being loaded from presets. :heavy_check_mark:
- Migrations for main prompt, nsfw, jailbreak are executed once when a preset is loaded. :heavy_check_mark:
- Character names are added when option is selected :heavy_check_mark:
 - Invalid character names cause warnings :heavy_check_mark:
- When utility prompts change, token usage is recalculated  :heavy_check_mark:
- Impersonation is working :heavy_check_mark:
- World Info format template is added :heavy_check_mark:
- NSFW avoidance prompt is added :heavy_check_mark:
- New chat prompt is added :heavy_check_mark:
- New group chat prompt is added in group chats :heavy_check_mark:
- New example chat prompt is added :heavy_check_mark:
- Continue nudge prompt is added when requested:heavy_check_mark:
- Replace empty message is added when input is empty :heavy_check_mark:

**Prompt Manager**

- Context size change triggers token recalculation :heavy_check_mark:
- Response length change triggers token recalculation :heavy_check_mark:
- Prompts can be created :heavy_check_mark:
- Prompts can be edited :heavy_check_mark:
- Prompts can be appended :heavy_check_mark:
- Prompts can be toggled :heavy_check_mark:
- Prompts can be detached :heavy_check_mark:
- Prompts can be deleted :heavy_check_mark:
- Prompts can be deleted while active, removing both prompt and attachment :heavy_check_mark:
- User prompts can be exported :heavy_check_mark:
- User prompts can be imported :heavy_check_mark:
- Prompts and their order can be reset for a character :heavy_check_mark:
- Prompts for a character can be exported with their order :heavy_check_mark:
- Prompts for a character can be imported with their order :heavy_check_mark:
- Markers can be inspected to list their contents.  :heavy_check_mark:

**Character**

- Editing main prompt override triggers recalculation :heavy_check_mark:
- Main prompt override is respected :heavy_check_mark:
- Editing the jailbreak override triggers recalculation :heavy_check_mark:
- Jailbreak prompt override is respected :heavy_check_mark:
- Creating a character causes a prompt manager entry to render :heavy_check_mark:
- Deleting a character causes the prompt manager entry to be removed :heavy_check_mark:
- Character prompt order reset :heavy_check_mark:

**Group Chat**

- Characters take turns according to their settings
- Character names are assigned properly to prompts if 'Add character names' is selected :heavy_check_mark:
- Groupchat specific main prompt is used :heavy_check_mark:
- Characters can be deleted :heavy_check_mark:
- Group chat can be edited :heavy_check_mark:
- Group chat can be deleted :heavy_check_mark:

**Authors Note**

- Authors not being sent after scenario :heavy_check_mark:
- Authors note being sent depending on depth :heavy_check_mark:
- Authors note character overrides affecting authors note :heavy_check_mark:

**Tavern Extras**

- Tavern Extras need to be tested. :exclamation:

Commit:
Add openai prompt manager edit popup

Commit:
Update openai default settings configuration file

Commit:
Add draggable list module

Used by openai prompt manager

Commit:
Add prompt manager module

Commit:
Add prompt manager configuration and default configuration

Commit:
Add prompt manager setup

Commit:
Let prompt manager control messages sent to openai

Commit:
Allow token counts to be cached when no character is selected

Commit:
Remove obsolete prompt related assignments

Commit:
Check oai settings for empty prompt manager configuration

Commit:
Add prompt manager styling

Commit:
Import promptmanager styles

Commit:
Add --white20a css variable

Commit:
Use saveSettings function instead of saveSettingsDebounced

Commit:
Add fluent getters to Chat Completion

Commit:
Enable Chat Completion to create messages for itself

And return their token count

Commit:
Add ability to update multiple prompts at once

To Promptmanager

Commit:
Remove obsolete sanitizing of prompts

Commit:
Show calculated prompts for makers

Commit:
Improve styling of Prompt Manager

Commit:
Rework of prepareOpenAIMessages

Commit:
Remove obsolete settings to update after change

Commit:
Fix quiet prompt functionality after rebasing

Commit:
Simplify impersonate prompt handling

Commit:
Never use token handler for openai

For now.

Commit:
Do not apped start new chat messages on example chat handling

The prompt manager does this job now.

Commit:
Update stylings for prompt manager prompts

Commit:
Clarify prompt managers total token count meaning

Commit:
Add marker symbol to prompt manager marker list entries

Commit:
Add error handling to parseTokenCounts function

Commit:
Remove obsolete properties on prompts

These are now handled by prompts_list

Commit:
Let oai prompt manager handle enhanced definitions

Commit:
Remove obsolete default settings in prompt manager constructor

Commit:
Disable role select only when advanced mode is off

Commit:
Add help note for prompt manager

Commit:
Refactor chat completion

Commit:
Add support for nsfw avoidance prompt

Commit:
Remove obsolete description of prompt manager

Commit:
Add missing export for missing identifier exception

Commit:
Remove obsolete jailbreak toggle.

Commit:
Simplify new character handling

Commit:
Improve documentation for prompt manager

Commit:
Refactor render functions of prompt manager

Improve readability and extendability

Commit:
Remove prompt list when character gets deleted

Commit:
Update documentation and code style for prompt manager

Commit:
Fix default prompts not being set correctly

Commit:
Remove obsolete interface elements and handling

Jailbreak etc.

Commit:
Fix dropspot highlight not appearing on hover

Commit:
Improve strength of main and impersonation prompt

Commit:
Show character definition in regular mode

Commit:
Add prompt manager support for group chats

Commit:
Prevent duplicate messages

By removing obsolete example message assignments

Commit:
Fire event when selecting a group chat

Commit:
Use EventEmitter instead of native event handling

Commit:
Update default jailbreak message

Commit:
Add i18n attributes to prompt manager elements

Commit:
Introduce new substitute parameters

charIfNotGroup and group

Commit:
Send main prompt with group chat

Cohee1207 committed:
Fix prompt settings init

Cohee1207 committed:
Fix sanitation again

Cohee1207 committed:
Hotfix

Cohee1207 committed:
Hotfix

Cohee1207 committed:
Hotfix

Cohee1207 committed:
Don't duplicate start chat marker if character has no example dialogues

Cohee1207 committed:
Reset token counters when switching characters. Put main prompt at the beginning

Commit:
Allow generate to run without sending a request

Commit:
Streamline token counting

By using TokenHandler instead of custom token handling

Commit:
Streamline token counting

By using TokenHandler instead of custom token handling

Commit:
Fire event after a chat has successfully loaded

Commit:
Update prompt manager after chat loaded

Commit:
Let prompt manager handle character and scenario prompts

Commit:
Enable TokenHandler to return total token count

Commit:
Fix extension prompt being not parsed correctly

Commit:
Add support for authors note and summary extensions

Commit:
Simplify example message handling

Commit:
Add prototype for prompt manager token management

Commit:
Introduction of additional helper classes, refactoring

Commit:
Add dialogue examples to chat completion

Commit:
Respect impersonate option

Commit:
Add nsfw avoidance prompt, quiet prompt and bias to chat completion

Commit:
Improve in-code documentation

Commit:
Fix chat completion not recognizing 0-position

Commit:
Add enhanced definitions

Commit:
Add memory and authors note to chat completion

Commit:
Refactor addMessageToChatCompletion

Commit:
Fix missing element terminator

Commit:
Handle various error scenarios during token counting

Commit:
Add dialogue examples first if user decides

Commit:
Apply character specific prompt overrides

Jailbreak and main

Commit:
Various improvements for stability and token counting.

Commit:
Fix prepared prompt not being returned

Commit:
Filter faulty prompt entries

Commit:
Disable toggling of main prompt

Commit:
Add event system to openai script

Lost during rebasing

Commit:
Do not count messages with empty prompts

Commit:
Visual improvements for mobile

Commit:
Support original-placeholder

Commit:
Declare prepareOpenAIMessages as synchronous

As there is no need for the function o be async anymore.

Commit:
Declare prepareOpenAIMessages as synchronous

As there is no need for the function o be async anymore.

Commit:
Render prompt manager after swiping with updated tokens

Including world info and extension prompts

Commit:
Restore default settings from dev branch

Commit:
Remove main prompt input field.

Being replaced by prompt manager

Commit:
Correct identifier used for jailbreak prompt override

Commit:
Add basic claude support

Commit:
Pass chat message name to chatcompletion endpoint

Commit:
Add basic WindowAI support

Commit:
Pass nsfwAvoidance message instead of prompt

Fixing invalid argument exception being thrown

Commit:
Fix prompt edit popup not appearing after refactoring

Commit:
Only add name to prompts when conforming to oai api standards

Commit:
Allow the user to toggle names in chat history

Commit:
Remove doubled initialization of prompt manager

Commit:
Add reset for main,nsfw and jailbreak prompts

Commit:
Fix jailbreak prompt not being overwritten by character

Commit:
Add support for persona description positioning

Commit:
Replace obsolete color variable

Commit:
Improve in-code documentation

Commit:
Fix prompt-manager being rendered twice.

Commit:
Raise warning threshold for chatHistory tokens

Commit:
Simplify token threshold warnings

Commit:
Save prompt manager items to perset

Commit:
Add migration for legacy prompts

Commit:
Add event for modifying settings before modifications

Commit:
Apply minor code style and stability improvements

Commit:
Remove faulty characters in default configuration

Commit:
Improve ui wording

Commit:
Throw a recoverable error on invalid character naming

Commit:
Add logging and minor stability enhancements

Commit:
Add simple profiling for token counting and rendering

Commit:
Fix handleSavePrompt logging throwing error

Commit:
Change draggable hover color to be more in tune with the default styling

Commit:
Add logging for prompt manager position changes

Commit:
Add fallback for empty dialogue example prompts

Commit:
Allow markers to inspected for their content

Commit:
Improve error messages

Commit:
Add inspection styles

Commit:
Create a deep copy of prompt defaults on char select

Along other minor fixes and optimizations

Commit:
Add support for empty prompt replacement

Commit:
Re-Render prompt manager on worldinfo settings change

Commit:
Add error handling to openai message counting

Commit:
Allow prompt manager list to grow in height

Commit:
Prevent chat history warnings from showing too early

Commit:
Don't use debounce for prompt manager save settings function

Commit:
Add support for cycle prompt

Commit:
Improve readability and clarity of prompt manager icons

Commit:
Add prompt export

Commit:
Refine export function

Commit:
Add confirmation step and allow character prompt order import

Commit:
Reintroduce event types after rebase

Commit:
Update prompt manager active character on character edit.

Commit:
Do not assign openai_max_tokens on input field changes

Fixes prompt manager falsely assigning string value to openai config

Commit:
Add info icons to prompt export

Commit:
Add icon for global prompts

Commit:
Always allow dropping prompts

Commit:
Do not render prompt export dialogue if not necessary

Commit:
Add documentation for chat completion

And dependant classes

Commit:
Fill tokenizer legacy counts where possible.

Commit:
Improve styling of inspect popup drawer

Commit:
Allow user to edit new chat, new group chat and new example chat prompts

Commit:
Add more padding in left panel scrollable

Commit:
Add margin utility classes

Commit:
Move utility prompts in its own, independent dropdown

And merge with existing advanced prompt bits

Commit:
Remove obsolete logs and improve documentation

Commit:
Fix continue functionality not working as expected

Commit:
Add continue nudge to editable utility prompts

Commit:
Do not attempt a dryrun if no character has been selected

Commit:
Validate import data before prompt import

Commit:
Add charIfNotGroup placeholder to default main prompt

Commit:
Do not generate group responses on dry run

Commit:
Replace DraggableList with jQuery UI sortable

Similar to Kobold AIs settings. Reducing code that needs maintaining.

Commit:
Add reset button for character prompt list

Commit:
Fix character reset not available in basic mode

Commit:
Fix invalid character name error not being handled

Commit:
Do not send empty scenario and personality prompts

Commit:
Fix authors note position not being considered

Commit:
Better visibility of prompt manager popup on mobile

Commit:
Prevent prompt manager initialization more than once

Commit:
Refactor documentation and make code more readable

No functional changes

Commit:
Add reset prompt support for enhance definitions

Commit:
Do not show reset button for user prompts

Commit:
When a prompt is reset, also reset name and role

Commit:
Initialize prompt manager on selecting openrouter, scale

Commit:
Refactor prepareOpenAiMessages, move prompt preparation in its own function

Commit:
Move main and jailbreak override from manager into openai script.

Makes the code clearer and easier to understand

Commit:
Use configured prefix for prompt manager classes instead of constant

Commit:
Refactor naming scheme for default promptmanager defaults

Commit:
Refactor token counting after completion. Simplify code.

Commit:
Refactor token counting after completion. Simplify code.

Commit:
Remove obsolete function getGroupMembers

Commit:
Refactor and simplify setup method for prompt manager.

aetherwu commented:
Great to hear this PR. I was just thinking about a similar strategy. 

Before every round that chat begins, a second thread would check the preferences:
- Should a specific character should care about the envirment and react to it? (One who doesn't like aircondition on, just for example. Or one always dislike icecream.)
- Then it feeds back the reaction (or none) back to main thread, where it would be composited in queue.
- Since it has spacial context to store personalized prompt, we can add a lot preferences into this list.

Commit:
Rename prompt lists to prompt order

Includes renaming of configuration, no functional changes

Commit:
Add user icon for user prompts

To be consistent with global and marker icons

Commit:
Set utility prompt text area size to 3 rows

Commit:
Move inspect functionality to prompt name click

Commit:
Fix edit button being shown for chat history and examples

Commit:
Fix prompt name not being greyed out when prompt disabled

Commit:
Add missing in-code documentation

Commit:
Fix in-code documentation typo

Commit:
Center globe icon

Commit:
Update existing prompts during migration

Instead of adding new prompts for main, jailbreak and nsfw

Commit:
Remove advanced options toggle

Commit:
Fix prompt manager not being draggable on mobile

Updates jquery ui touch punch with a maintained version, fixes other instances of jquery ui sortable as well (kobold, ...)

Commit:
Add theme tint and blur to prompt manager popup drawer

Commit:
Improve wording for prompt insertion

Commit:
Add quick edit drawer and fields for main and jailbreak

Commit:
Reduce font size on mobile

Commit:
Move quick edits on top of prompt manager

Commit:
Close and clear popup on preset change

Commit:
Remove obsolete characterSelected event

Cohee1207 committed:
Expand quick edits rows

Cohee1207 committed:
Sanitize completion names. Only send names in msg texts if names in completion is disabled

Cohee1207 committed:
#870 Fix Claude in prompt manager

Commit:
Append quietPrompt last

Commit:
Do not insert authors note into main

Since it is already present in the chat history

Commit:
Refactor addItem method of MessageCollection

Commit:
Emit event when a single chat message is deleted

Commit:
Render prompt manager when chat history is altered

Recounts tokens

Commit:
Emit events on changing chatcompletion source or model

Commit:
Render promptmanager when chatcompletion model or source change

Cohee1207 committed:
Use tokenizer-specific token cache

Commit:
Always add impersonate prompt last

Commit:
Swap detach and edit icons

Cohee1207 committed:
Minor spelling mistake

Cohee1207 committed:
Enable JB prompt by default

Cohee1207 committed:
Move assistant prefill to quick edit

Commit:
Change icon for global prompts

Commit:
Fix user messages not being added when populating example dialogue

Commit:
Remove obsolete debugging output

Commit:
Allow inspection for all prompts

Because, really, why not.

Commit:
Add message role to inspect

Commit:
Add flag to preset saving, allowing to prevent triggering the UI

Commit:
Fix preset migration

Commit:
Prevent preset migration from running multiple times.

artisticMink commented:
Don't merge please. There are still issues with preset migration.

Commit:
Copy migrated preset into loaded presets

Since this only is populated once during page load

Commit:
Remove obsolete assignment

Commit:
Use chatcompletion log function for token budget

Commit:
Fix prompts object reference being shared

By deepcloning a copy

Commit:
Prevent new example chat message being added for no reason

Commit:
Show tokens when inspecting a prompt

Commit:
Remove obsolete prompt manager settings object

Commit:
Create a deep clone of the preset before applying

Prevents non-primitives from causing overlapping references

artisticMink commented:
Presets are working now. The branch can be merged. 

I made some changes to onSettingsPresetChange(...), most notably that there's now a deep clone created from the preset, which is then applied onto the settings object. 

This should not affect any existing preset functionality as only primitives were applied and those were copied already. The clone prevents reference issues when applying non-primitive preset values.

Cohee1207 committed:
Enable nsfw prompt by default
